// @nrwl/react/src/utils/testing-generators.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createLib = exports.createWebApp = exports.createApp = void 0;
const tslib_1 = require("tslib");
const devkit_1 = require("@nrwl/devkit");
const application_1 = require("../generators/application/application");
const linter_1 = require("@nrwl/linter");
const web_1 = require("@nrwl/web");
function createApp(tree, appName) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const { fileName } = devkit_1.names(appName);
        yield application_1.default(tree, {
            babelJest: true,
            e2eTestRunner: 'none',
            linter: linter_1.Linter.EsLint,
            skipFormat: true,
            style: 'css',
            unitTestRunner: 'none',
            name: appName,
        });
    });
}
exports.createApp = createApp;
function createWebApp(tree, appName) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const { fileName } = devkit_1.names(appName);
        yield web_1.applicationGenerator(tree, {
            name: appName,
            skipFormat: true,
        });
    });
}
exports.createWebApp = createWebApp;
function createLib(tree, libName) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const { fileName } = devkit_1.names(libName);
        tree.write(`/libs/${fileName}/src/index.ts`, `import React from 'react';\n`);
        devkit_1.addProjectConfiguration(tree, fileName, {
            tags: [],
            root: `libs/${fileName}`,
            projectType: 'library',
            sourceRoot: `libs/${fileName}/src`,
            targets: {},
        });
    });
}
exports.createLib = createLib;
