// @nrwl/react/src/utils/versions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.babelPluginStyledComponentsVersion = exports.eslintPluginReactHooksVersion = exports.eslintPluginReactVersion = exports.eslintPluginJsxA11yVersion = exports.eslintPluginImportVersion = exports.typesReactReduxVersion = exports.reactReduxVersion = exports.reduxjsToolkitVersion = exports.testingLibraryReactVersion = exports.typesReactRouterDomVersion = exports.reactRouterDomVersion = exports.typesStyledJsxVersion = exports.styledJsxVersion = exports.emotionBabelPresetCssPropVersion = exports.emotionReactVersion = exports.emotionStyledVersion = exports.typesReactIsVersion = exports.reactIsVersion = exports.typesStyledComponentsVersion = exports.styledComponentsVersion = exports.typesReactDomVersion = exports.typesReactVersion = exports.reactDomVersion = exports.reactVersion = exports.nxVersion = void 0;
exports.nxVersion = '11.5.1';
exports.reactVersion = '17.0.1';
exports.reactDomVersion = '17.0.1';
exports.typesReactVersion = '16.9.56';
exports.typesReactDomVersion = '16.9.9';
exports.styledComponentsVersion = '5.2.1';
exports.typesStyledComponentsVersion = '5.1.4';
exports.reactIsVersion = '17.0.1';
exports.typesReactIsVersion = '16.7.1';
exports.emotionStyledVersion = '11.0.0';
exports.emotionReactVersion = '11.0.0';
exports.emotionBabelPresetCssPropVersion = '11.0.0';
exports.styledJsxVersion = '3.3.1';
exports.typesStyledJsxVersion = '2.2.8';
exports.reactRouterDomVersion = '5.2.0';
exports.typesReactRouterDomVersion = '5.1.6';
exports.testingLibraryReactVersion = '11.1.2';
exports.reduxjsToolkitVersion = '1.4.0';
exports.reactReduxVersion = '7.2.2';
exports.typesReactReduxVersion = '7.1.11';
exports.eslintPluginImportVersion = '2.22.1';
exports.eslintPluginJsxA11yVersion = '6.4.1';
exports.eslintPluginReactVersion = '7.21.5';
exports.eslintPluginReactHooksVersion = '4.2.0';
exports.babelPluginStyledComponentsVersion = '1.10.7';
