// @nrwl/jest/src/generators/jest-project/lib/update-workspace.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateWorkspace = void 0;
const devkit_1 = require("@nrwl/devkit");
function updateWorkspace(tree, options) {
    var _a;
    const projectConfig = devkit_1.readProjectConfiguration(tree, options.project);
    projectConfig.targets.test = {
        executor: '@nrwl/jest:jest',
        outputs: [
            devkit_1.joinPathFragments(devkit_1.normalizePath('coverage'), devkit_1.normalizePath(projectConfig.root)),
        ],
        options: {
            jestConfig: devkit_1.joinPathFragments(devkit_1.normalizePath(projectConfig.root), 'jest.config.js'),
            passWithNoTests: true,
        },
    };
    const isUsingTSLint = ((_a = projectConfig.targets.lint) === null || _a === void 0 ? void 0 : _a.executor) ===
        '@angular-devkit/build-angular:tslint';
    if (isUsingTSLint) {
        projectConfig.targets.lint.options.tsConfig = [
            ...(projectConfig.targets.lint.options.tsConfig || []),
            devkit_1.joinPathFragments(devkit_1.normalizePath(projectConfig.root), 'tsconfig.spec.json'),
        ];
    }
    devkit_1.updateProjectConfiguration(tree, options.project, projectConfig);
}
exports.updateWorkspace = updateWorkspace;
