// @nrwl/storybook/src/generators/cypress-project/cypress-project.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cypressProjectSchematic = exports.cypressProjectGenerator = void 0;
const tslib_1 = require("tslib");
const devkit_1 = require("@nrwl/devkit");
const cypress_1 = require("@nrwl/cypress");
const utilities_1 = require("../../utils/utilities");
function cypressProjectGenerator(tree, schema) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const e2eProjectName = schema.name + '-e2e';
        const installTask = yield cypress_1.cypressProjectGenerator(tree, {
            name: e2eProjectName,
            project: schema.name,
            js: schema.js,
            linter: schema.linter,
        });
        removeUnneededFiles(tree, e2eProjectName, schema.js);
        addBaseUrlToCypressConfig(tree, e2eProjectName);
        updateAngularJsonBuilder(tree, e2eProjectName, schema.name);
        yield devkit_1.formatFiles(tree);
        return installTask;
    });
}
exports.cypressProjectGenerator = cypressProjectGenerator;
function removeUnneededFiles(tree, projectName, js) {
    utilities_1.safeFileDelete(tree, devkit_1.readProjectConfiguration(tree, projectName).sourceRoot +
        (js ? '/integration/app.spec.js' : '/integration/app.spec.ts'));
    utilities_1.safeFileDelete(tree, devkit_1.readProjectConfiguration(tree, projectName).sourceRoot +
        (js ? '/support/app.po.js' : '/support/app.po.ts'));
}
function addBaseUrlToCypressConfig(tree, projectName) {
    const cypressConfigPath = devkit_1.readProjectConfiguration(tree, projectName).root + '/cypress.json';
    devkit_1.updateJson(tree, cypressConfigPath, (cypressConfig) => {
        cypressConfig.baseUrl = 'http://localhost:4400';
        return cypressConfig;
    });
}
function updateAngularJsonBuilder(tree, e2eProjectName, targetProjectName) {
    const project = devkit_1.readProjectConfiguration(tree, e2eProjectName);
    const e2eTarget = project.targets.e2e;
    project.targets.e2e = Object.assign(Object.assign({}, e2eTarget), { options: Object.assign(Object.assign({}, e2eTarget.options), { devServerTarget: `${targetProjectName}:storybook` }), configurations: {
            ci: {
                devServerTarget: `${targetProjectName}:storybook:ci`,
            },
        } });
    devkit_1.updateProjectConfiguration(tree, e2eProjectName, project);
}
exports.default = cypressProjectGenerator;
exports.cypressProjectSchematic = devkit_1.convertNxGenerator(cypressProjectGenerator);
