// @nrwl/workspace/src/command-line/print-affected.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.selectPrintAffected = exports.printAffected = void 0;
const tslib_1 = require("tslib");
const run_command_1 = require("../tasks-runner/run-command");
const utils_1 = require("../tasks-runner/utils");
const hasher_1 = require("../core/hasher/hasher");
const package_manager_1 = require("@nrwl/tao/src/shared/package-manager");
function printAffected(affectedProjectsWithTargetAndConfig, affectedProjects, projectGraph, { nxJson }, nxArgs, overrides) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const { runnerOptions } = run_command_1.getRunner(nxArgs, nxJson);
        const projectNames = affectedProjects.map((p) => p.name);
        const hasher = new hasher_1.Hasher(projectGraph, nxJson, runnerOptions);
        const tasksJson = yield createTasks(affectedProjectsWithTargetAndConfig, projectGraph, nxArgs, overrides, hasher);
        const result = {
            tasks: tasksJson,
            projects: projectNames,
            projectGraph: serializeProjectGraph(projectGraph),
        };
        if (nxArgs.select) {
            console.log(selectPrintAffected(result, nxArgs.select));
        }
        else {
            console.log(JSON.stringify(selectPrintAffected(result, null), null, 2));
        }
    });
}
exports.printAffected = printAffected;
function createTasks(affectedProjectsWithTargetAndConfig, projectGraph, nxArgs, overrides, hasher) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const tasks = affectedProjectsWithTargetAndConfig.map((affectedProject) => run_command_1.createTask({
            project: affectedProject,
            target: nxArgs.target,
            configuration: nxArgs.configuration,
            overrides,
            errorIfCannotFindConfiguration: false,
        }));
        const taskHashes = yield hasher.hashTasks(tasks);
        const pm = package_manager_1.detectPackageManager();
        const isYarn = pm === 'yarn';
        return tasks.map((task, index) => ({
            id: task.id,
            overrides,
            target: task.target,
            command: `${isYarn ? 'yarn' : `${pm} run`} ${utils_1.getCommandAsString('nx', isYarn, task)}`,
            outputs: utils_1.getOutputs(projectGraph.nodes, task),
            hash: taskHashes[index].value,
        }));
    });
}
function serializeProjectGraph(projectGraph) {
    const nodes = Object.values(projectGraph.nodes).map((n) => n.name);
    return { nodes, dependencies: projectGraph.dependencies };
}
function selectPrintAffected(wholeJson, wholeSelect) {
    if (!wholeSelect)
        return wholeJson;
    return _select(wholeJson, wholeSelect);
    function _select(json, select) {
        if (select.indexOf('.') > -1) {
            const [firstKey, ...restKeys] = select.split('.');
            const first = json[firstKey];
            throwIfEmpty(wholeSelect, first);
            const rest = restKeys.join('.');
            if (Array.isArray(first)) {
                return first.map((q) => _select(q, rest)).join(', ');
            }
            else {
                return _select(first, rest);
            }
        }
        else {
            const res = json[select];
            throwIfEmpty(wholeSelect, res);
            if (Array.isArray(res)) {
                return res.join(', ');
            }
            else {
                return res;
            }
        }
    }
}
exports.selectPrintAffected = selectPrintAffected;
function throwIfEmpty(select, value) {
    if (value === undefined) {
        throw new Error(`Cannot select '${select}' in the results of print-affected.`);
    }
}
