// @nrwl/workspace/src/generators/workspace/workspace.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.workspaceSchematic = exports.workspaceGenerator = exports.DEFAULT_NRWL_PRETTIER_CONFIG = void 0;
const tslib_1 = require("tslib");
const devkit_1 = require("@nrwl/devkit");
const versions_1 = require("../../utils/versions");
const fs_1 = require("fs");
const path_1 = require("path");
const workspace_1 = require("@nrwl/tao/src/shared/workspace");
exports.DEFAULT_NRWL_PRETTIER_CONFIG = {
    singleQuote: true,
};
function decorateAngularClI(host, options) {
    const decorateCli = fs_1.readFileSync(path_1.join(__dirname, '..', 'utils', 'decorate-angular-cli.js__tmpl__')).toString();
    host.write(path_1.join(options.directory, 'decorate-angular-cli.js'), decorateCli);
}
function setWorkspaceLayoutProperties(tree, options) {
    devkit_1.updateJson(tree, path_1.join(options.directory, 'nx.json'), (json) => {
        if (options.layout === 'packages') {
            json.workspaceLayout = {
                appsDir: 'packages',
                libsDir: 'packages',
            };
        }
        return json;
    });
}
function createAppsAndLibsFolders(host, options) {
    if (options.layout === 'packages') {
        host.write(path_1.join(options.directory, 'packages/.gitkeep'), '');
    }
    else {
        host.write(path_1.join(options.directory, 'apps/.gitkeep'), '');
        host.write(path_1.join(options.directory, 'libs/.gitkeep'), '');
    }
}
function createFiles(host, options) {
    const npmScope = options.npmScope ? options.npmScope : options.name;
    const formattedNames = devkit_1.names(options.name);
    devkit_1.generateFiles(host, path_1.join(__dirname, './files'), options.directory, Object.assign(Object.assign({ formattedNames, dot: '.', tmpl: '', workspaceFile: options.cli === 'angular' ? 'angular' : 'workspace', cliCommand: options.cli === 'angular' ? 'ng' : 'nx', nxCli: false, typescriptVersion: versions_1.typescriptVersion,
        prettierVersion: versions_1.prettierVersion,
        // angular cli is used only when workspace schematics is added to angular cli
        angularCliVersion: versions_1.angularCliVersion }, options), { nxVersion: versions_1.nxVersion,
        npmScope }));
}
function createPrettierrc(host, options) {
    devkit_1.writeJson(host, path_1.join(options.directory, '.prettierrc'), exports.DEFAULT_NRWL_PRETTIER_CONFIG);
}
function formatWorkspaceJson(host, options) {
    const path = path_1.join(options.directory, options.cli === 'angular' ? 'angular.json' : 'workspace.json');
    try {
        devkit_1.updateJson(host, path, (workspaceJson) => {
            const reformatted = workspace_1.reformattedWorkspaceJsonOrNull(workspaceJson);
            if (reformatted) {
                host.write(path, JSON.stringify(reformatted, null, 2));
            }
        });
    }
    catch (e) {
        console.error(`Failed to format: ${path}`);
        console.error(e);
    }
}
function workspaceGenerator(host, options) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        if (!options.name) {
            throw new Error(`Invalid options, "name" is required.`);
        }
        createFiles(host, options);
        createPrettierrc(host, options);
        if (options.cli === 'angular') {
            decorateAngularClI(host, options);
        }
        setWorkspaceLayoutProperties(host, options);
        createAppsAndLibsFolders(host, options);
        yield devkit_1.formatFiles(host);
        formatWorkspaceJson(host, options);
    });
}
exports.workspaceGenerator = workspaceGenerator;
exports.workspaceSchematic = devkit_1.convertNxGenerator(workspaceGenerator);
