// @nrwl/workspace/src/utilities/run-webpack.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getEmittedFiles = exports.runWebpackDevServer = exports.runWebpack = void 0;
const webpack = require("webpack");
const WebpackDevServer = require("webpack-dev-server");
const rxjs_1 = require("rxjs");
const path_1 = require("path");
const url = require("url");
function runWebpack(config) {
    return new rxjs_1.Observable((subscriber) => {
        const webpackCompiler = webpack(config);
        function callback(err, stats) {
            if (err) {
                subscriber.error(err);
            }
            subscriber.next(stats);
        }
        if (config.watch) {
            const watchOptions = config.watchOptions || {};
            const watching = webpackCompiler.watch(watchOptions, callback);
            return () => {
                watching.close(() => { });
            };
        }
        else {
            webpackCompiler.run((err, stats) => {
                callback(err, stats);
                subscriber.complete();
            });
        }
    });
}
exports.runWebpack = runWebpack;
function runWebpackDevServer(config) {
    return new rxjs_1.Observable((subscriber) => {
        var _a, _b;
        const webpackCompiler = webpack(config);
        let baseUrl;
        webpackCompiler.hooks.done.tap('build-webpack', (stats) => {
            subscriber.next({ stats, baseUrl });
        });
        const devServerConfig = config.devServer || {};
        const originalOnListen = devServerConfig.onListening;
        devServerConfig.onListening = function (server) {
            originalOnListen(server);
            const devServerOptions = server.options;
            baseUrl = url.format({
                protocol: devServerOptions.https ? 'https' : 'http',
                hostname: server.hostname,
                port: server.listeningApp.address().port,
                pathname: devServerOptions.publicPath,
            });
        };
        const webpackServer = new WebpackDevServer(webpackCompiler, devServerConfig);
        try {
            const server = webpackServer.listen((_a = devServerConfig.port) !== null && _a !== void 0 ? _a : 8080, (_b = devServerConfig.host) !== null && _b !== void 0 ? _b : 'localhost', function (err) {
                if (err) {
                    subscriber.error(err);
                }
            });
            return () => {
                server.close();
            };
        }
        catch (e) {
            throw new Error('Could not start start dev server');
        }
    });
}
exports.runWebpackDevServer = runWebpackDevServer;
function getEmittedFiles(stats) {
    const { compilation } = stats;
    const files = [];
    // adds all chunks to the list of emitted files such as lazy loaded modules
    for (const chunk of compilation.chunks) {
        for (const file of chunk.files) {
            files.push({
                // The id is guaranteed to exist at this point in the compilation process
                // tslint:disable-next-line: no-non-null-assertion
                id: chunk.id.toString(),
                name: chunk.name,
                file,
                extension: path_1.extname(file),
                initial: chunk.isOnlyInitial(),
            });
        }
    }
    // other all files
    for (const file of Object.keys(compilation.assets)) {
        files.push({
            file,
            extension: path_1.extname(file),
            initial: false,
            asset: true,
        });
    }
    // dedupe
    return files.filter(({ file, name }, index) => files.findIndex((f) => f.file === file && (!name || name === f.name)) ===
        index);
}
exports.getEmittedFiles = getEmittedFiles;
