// @nrwl/web/src/generators/init/init.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.webInitSchematic = exports.webInitGenerator = void 0;
const tslib_1 = require("tslib");
const devkit_1 = require("@nrwl/devkit");
const set_default_collection_1 = require("@nrwl/workspace/src/utilities/set-default-collection");
const run_tasks_in_serial_1 = require("@nrwl/workspace/src/utilities/run-tasks-in-serial");
const versions_1 = require("../../utils/versions");
const cypress_1 = require("@nrwl/cypress");
const jest_1 = require("@nrwl/jest");
function updateDependencies(tree) {
    devkit_1.updateJson(tree, 'package.json', (json) => {
        delete json.dependencies['@nrwl/web'];
        return json;
    });
    return devkit_1.addDependenciesToPackageJson(tree, {
        'core-js': '^3.6.5',
        'document-register-element': versions_1.documentRegisterElementVersion,
        tslib: '^2.0.0',
    }, {
        '@nrwl/web': versions_1.nxVersion,
    });
}
function initRootBabelConfig(tree) {
    if (tree.exists('/babel.config.json') || tree.exists('/babel.config.js')) {
        return;
    }
    devkit_1.writeJson(tree, '/babel.config.json', {
        babelrcRoots: ['*'],
    });
}
function webInitGenerator(tree, schema) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        let tasks = [];
        set_default_collection_1.setDefaultCollection(tree, '@nrwl/web');
        if (!schema.unitTestRunner || schema.unitTestRunner === 'jest') {
            const jestTask = jest_1.jestInitGenerator(tree, {});
            tasks.push(jestTask);
        }
        if (!schema.e2eTestRunner || schema.e2eTestRunner === 'cypress') {
            const cypressTask = cypress_1.cypressInitGenerator(tree);
            tasks.push(cypressTask);
        }
        const installTask = updateDependencies(tree);
        tasks.push(installTask);
        initRootBabelConfig(tree);
        if (!schema.skipFormat) {
            yield devkit_1.formatFiles(tree);
        }
        return run_tasks_in_serial_1.runTasksInSerial(...tasks);
    });
}
exports.webInitGenerator = webInitGenerator;
exports.default = webInitGenerator;
exports.webInitSchematic = devkit_1.convertNxGenerator(webInitGenerator);
